### ISSUE TITLE
Investigation: Rasterization State

### ISSUE THREAD

**Issue by Richard-Yunchao** [gpuweb/gpuweb#137]
# Introduction
There are a few widely used features that can impact rasterization fixed-function stage, for example, back face culling and front/back facing definition (ccw or cw), depth clip or clamp, rasterization discard, triangle fill mode, wide lines, etc. D3D12 and Vulkan have a standalone descriptor called rasterization descriptor/state to config these parameters in its graphics pipeline descriptor/state, and Metal can also set these parameters via some functions. 

This document attempts to describe these features among three native graphics APIs, in order to discuss some missing features in WebGPU idl.

**This document was updated/corrected according to comments below.**

# Native APIs
## D3D12
D3D12 has a descriptor named [RASTERIZER_DESC](https://docs.microsoft.com/en-us/windows/desktop/api/d3d12/ns-d3d12-d3d12_rasterizer_desc). Its definition is listed as follow: 
```cpp
typedef struct D3D12_RASTERIZER_DESC {
  D3D12_FILL_MODE                       FillMode;
  D3D12_CULL_MODE                       CullMode;
  BOOL                                  FrontCounterClockwise;
  INT                                   DepthBias;
  FLOAT                                 DepthBiasClamp;
  FLOAT                                 SlopeScaledDepthBias;
  BOOL                                  DepthClipEnable;
  BOOL                                  MultisampleEnable;
  BOOL                                  AntialiasedLineEnable;
  UINT                                  ForcedSampleCount;
  D3D12_CONSERVATIVE_RASTERIZATION_MODE ConservativeRaster;
} D3D12_RASTERIZER_DESC;
```

### Triangle Fill Mode
**Triangle Fill Mode** is used to specify that we draw 1) a solid triangle, or 2) only line/wireframe of the triangle, or 3) only point/vertex of the triangle. The advantage of triangle fill mode is that some operations like back-face culling can be used upon triangles with line or point fill mode. But if you draw lines or points directly, these features don't take effect because they operate only on complete triangles. 
 
D3D12 has a parameter `FillMode` to config this feature. The options are `fill (solid)` and `line (wireframe)`, see the definition of [FILL_MODE](https://docs.microsoft.com/en-us/windows/desktop/api/d3d12/ne-d3d12-d3d12_fill_mode) below:
```cpp
typedef enum D3D12_FILL_MODE {
  D3D12_FILL_MODE_WIREFRAME,
  D3D12_FILL_MODE_SOLID
} ;
```

### Front Face and Cull Mode
**Front Face** is used to specify which direction a triangle is facing via its winding order. Typically it uses counterclockwise (ccw) or clockwise(cw) to determine its front face, and the other winding order(cw or ccw) is back face. Once the direction of a triangle is determined, **Cull Mode** is used to specify a particular direction that a triangle falls to the specified direction will be culled and discarded. The typical options are none, back face, and front face.

**Front Face**: D3D12 has a parameter `FrontCounterClockwise`  to config this feature. If this member is **TRUE**, a triangle will be considered front-facing if its vertices are counter-clockwise on the render target and considered back-facing if they are clockwise. If this parameter is **FALSE**, the opposite is true. 

**Cull Mode**: D3D12 has a parameter `CullMode` to cull and discard triangles whose facing fall into the specified facing. `CullMode` has a few enumerations: `none`, `front`, and `back`. They are grouped into [CULL_MODE](https://docs.microsoft.com/en-us/windows/desktop/api/d3d12/ne-d3d12-d3d12_cull_mode) and listed as follow.
```cpp
typedef enum D3D12_CULL_MODE {
  D3D12_CULL_MODE_NONE,
  D3D12_CULL_MODE_FRONT,
  D3D12_CULL_MODE_BACK
} ;
```    

### Depth Clip or Clamp
**Depth Clip** is used to clip the depth value (z value). If the x or y coordinate of a vertex is outside of the view volume, it will be clipped away by default. But whether we clip a vertex whose z coordinate (depth) is outside of the view volume (near plane and far plane for depth) depends on this parameter. If depth clip is enabled, that vertex will be clipped away. If it is disabled, that vertex will be retained and drawn. The latter can be used to fill holes in geometry that would be caused by clipping. It is also called **Depth Clamp**, because it looks like the near and far planes are ignored and depth values are clamped to infinite near and far planes.

D3D12 has a parameter `DepthClipEnable`, which can be true or false in order to enable or disable depth clip. 

### Rasterization Discard
Both Metal and Vulkan can discard rasterization, which will stop the current pipeline before rasterization. D3D12 can support **Stream Output**, by which you can retrieve the geometry information right before rasterization and discard rasterization and all other stages afterwards. If this feature is enabled, it will enable rasterization discard implicitly. 

However, **I didn't find any parameter to explicitly enable/disable rasterization discard in D3D**. Please correct me if I am wrong. 

**(Updated according the comments below) We can enable rasterization discard via empty pixel shader + disable depth/stencil test.**
 
### Depth Bias
When two primitives are rendered on top of each other or very close to it, their interpolated depth values may be the same or very close. Due to floating-point imprecision, this can result in depth testing producing inconsistent and implementation-dependent results. This visual artifact is known as **depth fighting**. Depth bias is used to alleviate this visual artifact by setting a programmable bias to force primitives to be offset toward or away from the viewer. 

D3D12 has 3 parameters including `DepthBias` and `DepthBiasClamp` and `SlopeScaledDepthBias` to enable this feature. 
 
### Other features
**Anti-alias**: D3D12 has a few parameters like `MultisampleEnable` and `AntialiasedLineEnable` and `ForcedSampleCount` to cover anti-alias. We have a standalone investigation about MSAA by Jiawei. I will not cover anti-alias here. 

**Conservative Raster**: The last parameter left in rasterizer descriptor in D3D12 is `ConservativeRaster`. It is not supported in Metal and Vulkan. 

## Metal
Unlike D3D12, Metal doesn't have a particular descriptor to cover these features. But it has some functions to configure them. Most functions are methods in [MTLRenderCommandEncoder](https://developer.apple.com/documentation/metal/mtlrendercommandencoder).  
### Triangle Fill Mode
Metal has a function [MTLRenderCommandEncoder::setTriangleFillMode](https://developer.apple.com/documentation/metal/mtlrendercommandencoder/1516029-settrianglefillmode). This function can support two different triangle fill modes: `line` and `fill`, just like D3D12.
### Front Face and Cull Mode
**Front Face**: Metal has a function [MTLRenderCommandEncoder::setFrontFacingWinding](https://developer.apple.com/documentation/metal/mtlrendercommandencoder/1515499-setfrontfacingwinding). It can specify `cw` or `ccw` winding as triangle front face. And the other winding (`ccw` or `cw`) is thought to be back face. This feature is the same as the feature in D3D12. 

**Cull Mode**: Metal has a function [MTLRenderCommandEncoder::setCullMode](https://developer.apple.com/documentation/metal/mtlrendercommandencoder/1515975-setcullmode) to specify which facing will be culled and discarded. It can support `none`, `front`, and `back`, just like D3D12.  
### Depth Clip or Clamp
Metal has a function [MTLRenderCommandEncoder::setDepthClipMode](https://developer.apple.com/documentation/metal/mtlrendercommandencoder/1516267-setdepthclipmode). It can support depth `clip` and depth `clamp`, just like D3D12. 
### Rasterization Discard
Metal can enable rasterization discard via [isRasterizationEnabled](https://developer.apple.com/documentation/metal/mtlrenderpipelinedescriptor/1514708-israsterizationenabled) in [MTLRenderPipelineDescriptor](https://developer.apple.com/documentation/metal/mtlrenderpipelinedescriptor). 
### Depth Bias
Metal can set all three parameters for depth bias via [MTLRenderCommandEncoder::setDepthBias](https://developer.apple.com/documentation/metal/mtlrendercommandencoder/1516269-setdepthbias), just like D3D12. 
### Other features
Metal has a function named [setVisibilityResultMode](https://developer.apple.com/documentation/metal/mtlrendercommandencoder/1515556-setvisibilityresultmode) in **MTLRenderCommandEncoder**. It is used to control how to monitor samples that pass the depth and stencil tests. **But it is not supported in D3D12 and Vulkan**. 

## Vulkan
Vulkan has a standalone state to describe rasterization in its graphics pipeline: [VkPipelineRasterizationStateCreateInfo](https://www.khronos.org/registry/vulkan/specs/1.1/html/vkspec.html#VkPipelineRasterizationStateCreateInfo). I'd like to list it here:
```cpp
typedef struct VkPipelineRasterizationStateCreateInfo {
    VkStructureType                            sType;
    const void*                                pNext;
    VkPipelineRasterizationStateCreateFlags    flags;
    VkBool32                                   depthClampEnable;
    VkBool32                                   rasterizerDiscardEnable;
    VkPolygonMode                              polygonMode;
    VkCullModeFlags                            cullMode;
    VkFrontFace                                frontFace;
    VkBool32                                   depthBiasEnable;
    float                                      depthBiasConstantFactor;
    float                                      depthBiasClamp;
    float                                      depthBiasSlopeFactor;
    float                                      lineWidth;
} VkPipelineRasterizationStateCreateInfo;
```
### Triangle Fill Mode
`polygonMode` in this state is used to describe triangle fill mode in Vulkan. its type is [VkPolygonMode](https://www.khronos.org/registry/vulkan/specs/1.1/html/vkspec.html#VkPolygonMode), which can support 4 different triangle fill modes: `fill`, `line`, and `point`. **But Metal and D3D12 can only support fill and line**. 

**(From Corentin's comments below) Triangle fill mode other than "fill" are an optional Vulkan feature gated by `fillModeNonSolid` that doesn't have wide support on mobile GPUs.**
### Front Face and Cull Mode
**Front Face**: Vulkan can specify triangle's front face via `frontFace`. Its type is [VkFrontFace](https://www.khronos.org/registry/vulkan/specs/1.1/html/vkspec.html#VkFrontFace). This enumeration includes `ccw` and `cw`, just like Metal and D3D12. 

**Cull Mode**: Vulkan can cull and discard some particular triangles if its facing falls to the facing specified by 'cullMode'. Its type is [VkCullModeFlagBits](https://www.khronos.org/registry/vulkan/specs/1.1/html/vkspec.html#VkCullModeFlagBits), which includes `none`, `front`, `back`, and `front and back`. **But Metal and D3D12 can't support `front and back`.
### Depth Clip or Clamp
Vulkan can enable depth clip or depth clamp via `depthClampEnable`. If it is false, then depth clip is enabled. Otherwise, depth clamp is enabled. This feature is the same as the feature in Metal and D3D12.

**(From Corentin's comments below) Depth clamp is an optional Vulkan feature gated by `depthClamp` that's not widely supported on mobile chips.** 
### Rasterization Discard
Vulkan can enable rasterization discard via `rasterizationDiscardEnable`, just like Metal. 
### Depth Bias
Vulkan can enable depth bias via 4 parameters: `depthBiasEnable`, `depthBiasConstantFactor`, `depthBiasClamp`, and `depthBiasSlopeFactor`. While Metal and D3D12 only have the latter 3 parameters. But if these parameters are 0 in Metal and D3D12, then depth bias is disabled. 
### Other features
Metal has a parameter `lineWidth` to support wide lines. But Metal and D3D12 don't have this feature. Furthermore, some Vulkan drivers only support wide lines via sw like drawing some triangles to emulate wide lines. 
# Conclusion and Proposal
### Triangle Fill Mode
Metal, D3D12 can support 2 triangle fill modes: `fill`, and `line`. Vulkan may support one more fill mode `point`. But both `line` and `point` fill modes on Vulkan are optional. So we might not support this feature currently. 
### Front Face and Cull Mode
We can specify `ccw` or `cw` as front face in Metal, D3D12 and Vulkan. So we can support it in web idl.
Metal, D3D12 and Vulkan can support 3 cull modes: `none`, `front`, and `back`. So we can support these 3 cull modes in web idl. Vulkan can support one more cull mode `front and back`, though. 
### Depth Clip or Clamp
Metal and D3D12 can support depth clip/clamp. But depth clamp is an optional feature in Vulkan. However, dept clamp is useful for some scenarios. So we may support this feature as an extension. 
### Rasterization Discard
Metal and Vulkan can support rasterization discard directly. And it is doable in D3D12 because D3D12 can support stream output stage and enable rasterization discard (via an empty pixel shader + disable depth/stencil test). 

**(Updated from the comments below) But currently we can't guarantee that we can implement the technique like transform feedback or stream output in WebGPU (say via ssbo in vertex shader), so we might not to support rasterization discard in web idl right now.**
### Depth Bias
This feature can be supported in Metal, D3D12 and Vulkan in the same manner. **However, this feature is hardware-dependent, I think we might not expose this feature to WebGPU. But I'd like to hear your thoughts about it**. 

**(Updated from the comments below) Depth bias is used in some algorithms and requested by developers, so we will support it.** 
## Proposal
Since many features related to rasterization can be supported by all 3 native APIs, I suggest we support them in WebGPU idl. In addition, both D3D12 and Vulkan have a standalone state/descriptor to group these features together, So I'd like to have a descriptor name WebGPURasterizationStateDescriptor and a state named WebGPURastierizationState as follow: 
```cpp
enum WebGPUTriangleFillMode {
    "fill",
    "line"
};

enum WebGPUFrontFace {
    "ccw",
    "cw"
};

enum WebGPUCullMode {
    "none",
    "front",
    "back"
};

enum WebGPUDepthClipMode {
    "clip",
    "clamp"
};

dictionary WebGPURasterizationStateDescriptor {
    WebGPUTriangleFillMode fillMode;
    WebGPUFrontFace frontFace;
    WebGPUCullMode cullMode;
    WebGPUDepthClipMode clipMode;
    bool rasterizationDiscardEnable;
};

interface WebGPURasterizationState {
};
```

In addition, we need to add an instance of WebGPURasterizationState into **WebGPURenderPipelineDescriptor**: 
```cpp
WebGPURasterizationState rasterizationState;
```

Finally, we need to add a function into **WebGPUDevice**: 
```cpp
WebGPURasterizationState createRasterizationState(WebGPURasterizationStateDescriptor descriptor); 
```

# TODO
Update the WebGPU idl once we get consensus on this.

**Comment by Richard-Yunchao** [comment]
PTAL when you have time, @Kangz , @kainino0x , @kvark , @RafaelCintron , @litherum. I'd like to hear your thoughts about these features. I don't know whether you discussed about these features before, but I can't find any discussion at https://github.com/gpuweb/gpuweb/blob/master/design/Pipelines.md.

**Comment by Kangz** [comment]
- `fillMode`: modes other than "fill" are an optional Vulkan feature gated by `fillModeNonSolid` [that doesn't have wide support on mobile GPUs](https://vulkan.gpuinfo.org/listreports.php?feature=fillModeNonSolid&option=not). I suggest we drop it.
 - `depthClamp` is also an optional Vulkan feature that's not support on mobile chips so we should drop it.
 - `rasterizationDiscardEnable`: for D3D12 I believe rasterizer discarding is done by not specifying a pixel shader. (at least that's what ANGLE does on D3D11). That said I'm not sure how useful rasterization discard will be given that WebGPU won't be able to guarantee vertex shaders can use SSBOs (and will have no transform feedback).
 - `frontFace` and `cullMode` LGTM. Note that Vulkan's cull mode "front and back" corresponds to D3D12's and Metal's "none" where no triangles are culled. This doesn't change the proposal.
 - Depth Bias and friends: I believe this will be a very requested feature because it is very often used in shadow map algorithms.

Also the proposal should be updated in the style of #134

**Comment by magcius** [comment]
I think depth clamp would be helpful to have as an optional extension. It's very, very common for shadow maps as well.

**Comment by Richard-Yunchao** [comment]
> * `fillMode`: modes other than "fill" are an optional Vulkan feature gated by `fillModeNonSolid` [that doesn't have wide support on mobile GPUs](https://vulkan.gpuinfo.org/listreports.php?feature=fillModeNonSolid&option=not). I suggest we drop it.
> * `depthClamp` is also an optional Vulkan feature that's not support on mobile chips so we should drop it.

Whoops, I didn't notice that these two features are optional in Vulkan, while they must are required in D3D12 and Metal (iOS 11+). 
> * `rasterizationDiscardEnable`: for D3D12 I believe rasterizer discarding is done by not specifying a pixel shader. (at least that's what ANGLE does on D3D11). That said I'm not sure how useful rasterization discard will be given that WebGPU won't be able to guarantee vertex shaders can use SSBOs (and will have no transform feedback).

Yes, I also looked into ANGLE code, and it disabled pixel shader if we enable rasterization discard. In addition, I agree that if feature like transform feedback (stream output in D3D12) is not supported in WebGPU, rasterization discard might be not useful. 
> * `frontFace` and `cullMode` LGTM. Note that Vulkan's cull mode "front and back" corresponds to D3D12's and Metal's "none" where no triangles are culled. This doesn't change the proposal.

Hmm, I checked Vulkan spec again, its `front an back` cull mode do cull both front facing and back facing and all triangles are discarded. It is different from `none` in D3D12 and Metal. And Vulkan do have `none` cull mode, which is equivalent to `none` cull mode in D3D12 and Metal. 

`front and back` cull mode will cull all triangles. I don't think it is very useful. It might be the reason that D3D12 and Metal don't support `front and back` cull mode. 
> * Depth Bias and friends: I believe this will be a very requested feature because it is very often used in shadow map algorithms.

Glad to know this, I will add depth bias related feature if needed.
> 
> Also the proposal should be updated in the style of #134

Got it. Thank you, @Kangz .

**Comment by Richard-Yunchao** [comment]
> > or D3D12 I believe rasterizer discarding is done by not specifying a pixel shader. (at least that's what ANGLE does on D3D11). That said I'm not sure how useful rasterization discard will be given that WebGPU won't be able to guarantee vertex shaders can use SSBOs (and will have no transform feedback)
> 
> Noo, you have an empty pixel shader for Z-prepass where you have no colors to output, just depth and you actually don't want to launch a pixel shader thread just to straight to early-fragment-test part of the pipeline and stop there.
> Rasterizer Discard stops the whole pixel generation from happening.
> Its a separate setting in itself.

So, do you know how to enable rasterization discard in D3D if it is a separate setting in itself and different from **enabling stream output + making pixel shader null**? Just curious about this from the technical point, even though rasterization discard will not be exposed in WebGPU idl. Thank you.

**Comment by Kangz** [comment]
> Hmm, I checked Vulkan spec again, see VkCullModeFlagBits. its front an back cull mode do cull both front facing and back facing and all triangles are discarded. It is different from none in D3D12 and Metal. You know, Vulkan do have none cull mode, which is equivalent to none cull mode in D3D12 and Metal.

Apologies, I had misread ANGLE's source code. You're correct.

**Comment by kvark** [comment]
I think we can proceed without the rasterizer discard for now.

**Comment by kvark** [comment]
@Richard-Yunchao thank you for the investigation!

> If this feature is enabled, it will enable rasterization discard implicitly.

I couldn't find it in the docs. Could you point me to where this is documented?

> Depth Bias
This feature can be supported in Metal, D3D12 and Vulkan in the same manner. However, this feature is hardware-dependent, I think we might not expose this feature to WebGPU. But I'd like to hear your thoughts about it.

I agree with @Kangz , this feature is a total must to support. The only caveat here is that D3D12 uses integer depth bias constant factor, while Vulkan and Metal accept floats. I think we should make it an integer as well to avoid troubles with portability.

> Metal has a function named setVisibilityResultMode in MTLRenderCommandEncoder. It is used to control how to monitor samples that pass the depth and stencil tests. But it is not supported in D3D12 and Vulkan.

This is covered by #72.

**Comment by magcius** [comment]
> Got any links? The only relevant piece I found (by searching quickly) is ARB_depth_clamp saying it's useful for stencil shadow volumes. This is hardly a "very, very common" use case.

Sure thing. It's known as "cascade shadow map pancaking". https://twitter.com/knarkowicz/status/665159371375120384

**Comment by devshgraphicsprogramming** [comment]
> Got any links? The only relevant piece I found (by searching quickly) is ARB_depth_clamp saying it's useful for stencil shadow volumes. This is hardly a "very, very common" use case.

That is actually still a very common use case, sometimes they are used to draw light volumes in Deferred Rendering (old engines).

But recently I used stencil shadow algorithm to render SDF shadows, and I'm certain that Unreal Engine 4 uses the same approach to build its raytraced SDF shadow buffer.

**Comment by devshgraphicsprogramming** [comment]
> I agree that if feature like transform feedback (stream output in D3D12) is not supported in WebGPU, rasterization discard might be not useful.

No you still need rasterizer discard, because transform feedback is not available anymore because you can implement it yourself far more efficiently with SSBO and atomic operations on SSBOs (DX11 feature).

P.S. This is also the reason why Atomic Counter Buffers are gone from Vulkan, because you can now do atomics on arbitrary addresses with SSBOs.

**Comment by Richard-Yunchao** [comment]
> > I agree that if feature like transform feedback (stream output in D3D12) is not supported in WebGPU, rasterization discard might be not useful.
> 
> No you still need rasterizer discard, because transform feedback is not available anymore because you can implement it yourself far more efficiently with SSBO and atomic operations on SSBOs (DX11 feature).
> 
> P.S. This is also the reason why Atomic Counter Buffers are gone from Vulkan, because you can now do atomics on arbitrary addresses with SSBOs.

Sorry, I am not clear here, @devshgraphicsprogramming. Transform feedback is not supported in Vulkan, In addition, just like @Kangz said, WebGPU won't be able to guarantee vertex shaders can use SSBOs. So developers can't implement transform feedback either. So it seems that we can't read geometry data back before rasterization, which means that rasterization discard might not be useful.  BTW, you talked that: rasterization discard is a separate setting in itself, it's different from making pixel shader empty in D3D. So. do you know how to enable rasterization discard in D3D? I am just curious about this from technical perspective.

**Comment by devshgraphicsprogramming** [comment]
> Transform feedback is not supported in Vulkan

This I was not contradicting in any way shape or form.

>  WebGPU won't be able to guarantee vertex shaders can use SSBOs

Thats kind of the reason why most people are using compute+SSBO atomics rather than vertex+rasterizer discard+SSBO.

But vertex shader has nice cache optimizations so it could be more efficient than compute for certain tasks.

> rasterization discard is a separate setting in itself, it's different from making pixel shader empty in D3D

From the DirectX doc:

> You may disable rasterization by telling the pipeline there is no pixel shader (set the pixel shader stage to NULL with ID3D11DeviceContext::PSSetShader), and disabling depth and stencil testing (set DepthEnable and StencilEnable to FALSE in D3D11_DEPTH_STENCIL_DESC). While disabled, rasterization-related pipeline counters will not update. There is also a complete description of the rasterization rules.

You need to both specify a null pixel shader, and both disable the depth and stencil tests... then you get a rasterizer discard equivalent.

Just a null pixel shader with z-tests and stencil tests will give you a z-prepass-like rasterization.

**Comment by Richard-Yunchao** [comment]
Thanks for your explanation, @devshgraphicsprogramming .

**Comment by litherum** [comment]
I somehow missed the part of the spec that talks about SSBOs in vertex shaders. Can you point me to a place where I can read about that restriction? Is there an extension to enable it? Is the extension widely supported?

**Comment by litherum** [comment]
This is a fantastic analysis, thanks for putting this together!

**Comment by Kangz** [comment]
@litherum it is `VkPhysicalDeviceFeatures::vertexPipelineStoresAndAtomics`.

**Comment by kvark** [comment]
Ugh, this is annoying. Samsung Galaxy phones, some Intel iGPUs, and a few others don't want vertex shaders to do stores into buffers/images.

**Comment by krogovin** [comment]
For Intel Hardware, if memory serves correctly, it is only Broadwell or older that won't do vertex shaders having stores into buffers/images, and that is only with the Mesa driver.

Gen10 GPU's, even with Mesa, should support stores into buffers/images for all shader stages. On paper (if the vertex shader is using the scalar shader compiler backend), stores to images/buffer should also work on. Starting in Gen8, there is an environmental variable that when set makes it possible, INTEL_SCALAR_VS. 

FWIW, the initial implementation of image-load-store on Mesa as far back as Gen7 (IvyBridge) *DID* implement it in all shader stages, but a few of theMesa community (read two or three Intel employees) objected to the implementer's (another Intel employee) use of templates and he needed to get the feature out the door so he dropped support in the vec4 compiler backend rather that write the feature twice.

**Comment by devshgraphicsprogramming** [comment]
I might point this out, although I'm sure everyone alreayd knows this, that even if hardware doesn't allow SSBO in vertex shaders, it would still allow Uniform Texel Buffers (Texture Buffer Objects in OpenGL) AFAIK.

**Comment by krogovin** [comment]
WebGPU does have read-only shader-buffer objects which are sort-of-ish the same as TBO's. The only ugly, but it actually is not that bad, is that the shader generation emitted by WebGPU will need to translate variable assess to texelFetch command appropriately.

This is not so bad, but the more components it fetches at once the better though.

**Comment by devshgraphicsprogramming** [comment]
TBO's in OpenGL require `texelFetch` as well.

**Comment by kvark** [comment]
Note that SSBOs are not only used for stores, they can also be used as "readonly" for loads in the vertex shaders, which isn't restricted by `VkPhysicalDeviceFeatures::vertexPipelineStoresAndAtomics`. So we shouldn't need to back "read-only buffer" bindings by TBOs.

**Comment by devshgraphicsprogramming** [comment]
>  they can also be used as "readonly" for loads in the vertex shaders, which isn't restricted by VkPhysicalDeviceFeatures::vertexPipelineStoresAndAtomics. 

As long as that is 100% an option then TBOs could be out of the picture, however TBOs have the advantage of automatic conversions and caching.

Vertex attributes can only be (d/u/i)vec(2,3,4) or (u)int/float/double, some formats are complex to decode like RG8_SNORM . TBO map very nicely to those.

**Comment by devshgraphicsprogramming** [comment]
There's one last thing not to overlook, on OpenGL (ES 3.0) and Vulkan, TB/TBO counts against the texture slot limit, not against the SSBO limit. So you can use faaar more TBOs than SSBOs.

**Comment by devshgraphicsprogramming** [comment]
There's one last thing to not overlook TBO/UBO count against the texture slot limit, not the SSBO limit, and those two limits are distinct. So you can use faaar more buffers as input since you can use some buffers as ssbo and some as tbo and the count in two slot limits sums (especially good for vertex pulling in clamping indices for robust buffer access, since you're not killing SSBO slots).

**Comment by magcius** [comment]
> Thats kind of the reason why most people are using compute+SSBO atomics rather than vertex+rasterizer discard+SSBO.
>
> But vertex shader has nice cache optimizations so it could be more efficient than compute for certain tasks.

I missed this the first time. The vertex cache is pretty much disabled when external stores happen, since it's observable.

**Comment by krogovin** [comment]
> I missed this the first time. The vertex cache is pretty much disabled when writing to SSBOs from vertex shaders, since it's observable.

I have memory that the spec for vertex shader with side-effects says that the number of times the vertex shader is called on a vertex is non-deterministic because of the use of a post-vertex cache, i.e. the spec gave room for implementation to leave the post-vertex cache.

On another note, as transform feed back was somewhat mentioned.... it is a crying shame that Vulkan only has transform feed back as an extension (or planned extension). It is not possible to emulate transform feedback (and get the ordering correct) with tessellation shaders, or geometry shaders with number of primitives not constant per invocation..  or even with just a vertex shader if primitive restart is active. Some folk often cite that many use cases for transform feedback can be done via compute, the truth is that even those use cases, some HW vendors will be faster with transform feedback than compute because of a few fixed function parts doing the work. Perhaps we can consider if transform feedback as an optional extension?

**Comment by devshgraphicsprogramming** [comment]
> It is not possible to emulate transform feedback (and get the ordering correct) with tessellation shaders, or geometry shaders with number of primitives not constant per invocation.. or even with just a vertex shader if primitive restart is active.

You just need to sort at the end (in packets of subgroup-size), or bucket per workgroup and compact at the end. If AMD any any other vendor actually gave us `ds_ordered_count` as an extension then we wouldn't have this problem.

> Some folk often cite that many use cases for transform feedback can be done via compute, the truth is that even those use cases, some HW vendors will be faster with transform feedback than compute because of a few fixed function parts doing the work. Perhaps we can consider if transform feedback as an optional extension?

Only on Intel and some mobile possibly, PC has already shifted away from that (AMD a long time ago, see `ds_ordered_count`). But most mobile does not have xform feedback, geometry shaders and tessellation.

**Comment by jdashg** [comment]
Is there any mobile subset that we're expecting to get WebGPU that doesn't have es32? (transform feedback was es30, geo and tess shaders are es32) Or is the point that these are not fixed-function, and therefore replaceable with compute?

**Comment by devshgraphicsprogramming** [comment]
>  Or is the point that these are not fixed-function, and therefore replaceable with compute?

Mostly that, and the fact that AMD already does transform feedback with an SSBO + atomic counter.

**Comment by Degerz** [comment]
> Mostly that, and the fact that AMD already does transform feedback with an SSBO + atomic counter.

I believe this only applies to open source drivers for VERY recent AMD hardware such as Navi. On older AMD HW like Vega or other drivers, they use a legacy fixed function path for transform feedbacks. I think every other desktop graphics vendor including Intel and Nvidia also uses fixed function hardware to accelerate transform feedbacks as well ... 

Emulating transform feedbacks via global ordered append with primitive shaders (NGG) is an exceptionally unique case to AMD Navi ... 

For many other vendors, I don't see compute being a good replacement to exposing fixed function hardware for transform feedbacks and that applies even further with complex cases like geometry shaders or tessellation ...

**Comment by Kangz** [comment]
Closing. We added the rasterization state a long time ago. Further changes can go in new issues.
